torchtext/prototype/generate.py

```diff
@@ -0,0 +1,324 @@
+                    model_inputs["return_dict"] = True
+                    model_inputs["output_hidden_states"] = True
+
+                outputs = self.model(**model_inputs)
+                output_key = "logits" if self.is_huggingface_model else "decoder_output"
+                lm_scores = outputs[output_key]
+
+                # Keep track of probabilities over vocab for this pairing
+                out_probs.append(torch.squeeze(lm_scores[:, -1]).tolist())
+                
+                # Keep track of sequence and decoder hidden states
+                model_states.append(
+                    create_emitting_model_state(
+                        Seq2SeqModelState(
+                            timestep=timestep,
+                            sequence=new_input_ids,
+                            lm_scores=lm_scores
+                        )
+                    )
+                )
+
+            return out_probs, model_states
+
+        options = LexiconFreeSeq2SeqDecoderOptions(
+            beam_size=num_beams,
+            beam_size_token=beam_size_token,
+            beam_threshold=beam_threshold,
+            lm_weight=0.0,  # We have no custom LM so score is zero
+            eos_score=eos_score,
+            log_add=True,
+        )
+
+        decoder = LexiconFreeSeq2SeqDecoder(
+            options=options,
+            lm=ZeroLM(),
+            eos_idx=eos_idx,
+            update_func=update_func,
+            max_output_length=max_len
+        )
+
+        # Create these as function b/c unnamed functions (lambdas) cause problems w/ MP
+        def select_second_elem_in_tuple(tup: Tuple[List[int], float]) -> float:
+            return tup[1]
+        
+        def is_not_neg_one(elem: int) -> bool:
+            return elem != -1
+
+        def beam_decode_step(timestep: int) -> torch.Tensor:
+            # Decode step takes ptr to encoder emissions, i, and beam size token
+            # but actually these aren't currently being used.
+            decoder.decode_step(0, timestep, 0)
+            hyps = decoder.get_all_final_hypothesis()
+
+            # Find the best beam
+            token_scores = [(hyp.tokens, hyp.score) for hyp in hyps]
+            final_tokens = list(
+                filter(
+                    is_not_neg_one, 
+                    max(token_scores, key=select_second_elem_in_tuple)[0]
+                )
+            )
+
+            # Hack, but have to prepend the input tokens if decoder-only model
+            if not self.is_encoder_decoder:
+                final_tokens = input_ids[timestep].tolist() + final_tokens
+
+            # Makeshift padding so that we can stack the tensors
+            while len(final_tokens) < max_len:
+                final_tokens += [0]
+            
+            # Convert from list to tensors
+            final_tokens_as_tensors = torch.Tensor(final_tokens).to(torch.long)
+
+            return final_tokens_as_tensors
+
+        if num_python_workers > 1:
```

Comment:
PyTorch has a MP module (essentially a clone of Python's MP); however, this relies on `pickle`, which means all of these functions would have to be at a global level. Very open to suggestions here.